File: src/main/java/com/osrs/dps/calc/EquipmentTotals.java
package com.osrs.dps.calc;

import com.osrs.dps.model.AttackType;
import com.osrs.dps.model.EquipmentItem;
import com.osrs.dps.model.Monster;
import com.osrs.dps.model.PlayerSetup;

/**
 * Aggregated equipment bonuses with the wiki tool's gear-level adjustments applied
 * (Tumeken's shadow tripling, Dinh's bulwark strength, Virtus + ancients,
 * elite void mage visible bonus, Keris partisan of amascut penalty outside ToA).
 */
final class EquipmentTotals {

    int str;
    int rangedStr;
    /** Magic damage bonus in tenths of a percent. */
    int magicStrTenths;
    int stab;
    int slash;
    int crush;
    int ranged;
    int magic;

    EquipmentTotals(PlayerSetup p, Monster m, Gear g) {
        for (EquipmentItem item : p.getEquipment().values()) {
            str += item.bonuses.str;
            rangedStr += item.bonuses.rangedStr;
            magicStrTenths += item.bonuses.magicStr;
            stab += item.offensive.stab;
            slash += item.offensive.slash;
            crush += item.offensive.crush;
            ranged += item.offensive.ranged;
            magic += item.offensive.magic;
        }

        // The shadow's bonus multiplier does not apply when manually casting a spell
        boolean manualCast = p.getStance() == com.osrs.dps.model.Stance.MANUAL_CAST;
        if (g.tumekensShadow() && !manualCast) {
            int factor = CombatConstants.TOA_MONSTER_IDS.contains(m.id) ? 4 : 3;
            magicStrTenths = Math.min(1000, magicStrTenths * factor);
            magic *= factor;
        }

        if ("Keris partisan of amascut".equals(g.weaponName())
                && !CombatConstants.TOA_MONSTER_IDS.contains(m.id)) {
            str -= 22;
            stab -= 50;
        }

        if (g.wearing("Dinh's bulwark", "Dinh's blazing bulwark")) {
            int defenceSum = 0;
            for (EquipmentItem item : p.getEquipment().values()) {
                defenceSum += item.defensive.stab + item.defensive.slash
                        + item.defensive.crush + item.defensive.ranged;
            }
            str += Math.max(0, (defenceSum - 800) / 12 - 38);
        }

        if (p.getSpell() != null && "ancient".equals(p.getSpell().spellbook)
                && p.isCastingSpell()) {
            magicStrTenths += 30 * g.virtusPieces();
        }

        if (g.eliteMagicVoid()) {
            magicStrTenths += 50;
        }
    }

    int attackBonus(AttackType type) {
        return switch (type) {
            case STAB -> stab;
            case SLASH -> slash;
            case CRUSH -> crush;
            case RANGED -> ranged;
            case MAGIC -> magic;
        };
    }

    /** Base max hit formula from effective level and a strength-style bonus. */
    static int maxHitFromEffective(int effectiveLevel, int strengthBonus) {
        return (effectiveLevel * (strengthBonus + 64) + 320) / 640;
    }
}

File: src/calc/equipmentTotals.ts
import type { AttackType, Monster, PlayerSetup } from '../model/types';
import { TOA_MONSTER_IDS } from './constants';
import type { Gear } from './gear';

export interface EquipmentTotals {
  str: number;
  rangedStr: number;
  /** Magic damage bonus in tenths of a percent */
  magicStrTenths: number;
  stab: number;
  slash: number;
  crush: number;
  ranged: number;
  magic: number;
}

/** Base max hit formula from effective level and a strength-style bonus. */
export const maxHitFromEffective = (effectiveLevel: number, strengthBonus: number): number =>
  Math.trunc((effectiveLevel * (strengthBonus + 64) + 320) / 640);

export const attackBonus = (t: EquipmentTotals, type: AttackType): number => t[type];

/**
 * Aggregated equipment bonuses with gear-level adjustments (Tumeken's shadow
 * tripling, Dinh's bulwark strength, Virtus + ancients, elite void mage,
 * Keris partisan of amascut penalty outside ToA).
 */
export const computeTotals = (p: PlayerSetup, m: Monster, g: Gear,
  spellbook: string | null): EquipmentTotals => {
  const t: EquipmentTotals = {
    str: 0, rangedStr: 0, magicStrTenths: 0,
    stab: 0, slash: 0, crush: 0, ranged: 0, magic: 0,
  };
  for (const item of Object.values(p.equipment)) {
    if (!item) continue;
    t.str += item.bonuses.str;
    t.rangedStr += item.bonuses.ranged_str;
    t.magicStrTenths += item.bonuses.magic_str;
    t.stab += item.offensive.stab;
    t.slash += item.offensive.slash;
    t.crush += item.offensive.crush;
    t.ranged += item.offensive.ranged;
    t.magic += item.offensive.magic;
  }

  // The shadow's multiplier does not apply when manually casting
  const manualCast = p.stance === 'Manual Cast';
  if (g.tumekensShadow() && !manualCast) {
    const factor = TOA_MONSTER_IDS.has(m.id) ? 4 : 3;
    t.magicStrTenths = Math.min(1000, t.magicStrTenths * factor);
    t.magic *= factor;
  }

  if (g.weaponName() === 'Keris partisan of amascut' && !TOA_MONSTER_IDS.has(m.id)) {
    t.str -= 22;
    t.stab -= 50;
  }

  if (g.wearing("Dinh's bulwark", "Dinh's blazing bulwark")) {
    let defenceSum = 0;
    for (const item of Object.values(p.equipment)) {
      if (!item) continue;
      defenceSum += item.defensive.stab + item.defensive.slash
        + item.defensive.crush + item.defensive.ranged;
    }
    t.str += Math.max(0, Math.trunc((defenceSum - 800) / 12) - 38);
  }

  const casting = p.attackType === 'magic'
    && (p.stance === 'Autocast' || p.stance === 'Defensive Autocast' || manualCast);
  if (spellbook === 'ancient' && casting) {
    t.magicStrTenths += 30 * g.virtusPieces();
  }

  if (g.eliteMagicVoid()) {
    t.magicStrTenths += 50;
  }

  return t;
};
